File: code_samples/section11/problems/problem11_5/problem11_5.js
class DSU {
    /*
     * DSU (Disjoint Set Union) / Union-Find
     *
     * Purpose:
     *   Maintain disjoint sets over elements {0..n-1} while supporting:
     *     - find(x):    get the representative (root) of x's set
     *     - unite(a,b): merge the sets containing a and b
     *
     * Key ideas:
     *   - parent[] forms a forest of trees; each set is one tree.
     *   - rank[] is a heuristic to keep trees shallow (union by rank).
     *   - path compression in find() flattens trees for faster future queries.
     *
     * Extra convenience:
     *   - count tracks how many disjoint sets currently exist.
     *     This makes "number of connected components" easy to compute.
     */
    constructor(n) {
        /*
         * Initialize n singleton sets:
         *   parent[i] = i means every node starts as its own root.
         *   rank[i] = 0 because all trees are size/height 1 initially.
         *   count = n because there are n separate components at the start.
         */
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = new Array(n).fill(0);
        this.count = n;
    }

    find(x) {
        /*
         * Find the root representative of x's set.
         *
         * Path compression:
         *   If x is not a root, recursively find the root and rewrite parent[x]
         *   to point directly to it. This drastically speeds up future finds.
         */
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    unite(a, b) {
        /*
         * Union/merge the sets that contain a and b.
         *
         * Steps:
         *   1) Find roots ra and rb.
         *   2) If they match, already in same set → nothing to do.
         *   3) Otherwise, attach the lower-rank root under the higher-rank root.
         *   4) If ranks are equal, pick one as new root and increment its rank.
         *   5) Decrement count exactly once for a successful merge (one fewer set).
         */
        let ra = this.find(a), rb = this.find(b);
        if (ra === rb) return;

        if (this.rank[ra] < this.rank[rb]) {
            this.parent[ra] = rb;
        } else if (this.rank[ra] > this.rank[rb]) {
            this.parent[rb] = ra;
        } else {
            this.parent[rb] = ra;
            this.rank[ra]++;
        }
        this.count--;
    }
}

// edges: array of [u, v]
function countComponents(n, edges) {
    /*
     * Count connected components in an undirected graph.
     *
     * Parameters:
     *   n     : number of nodes labeled 0..n-1
     *   edges : list of undirected edges, each as [u, v]
     *
     * Approach (DSU):
     *   - Start with n components (each node alone).
     *   - For each edge (u, v), union their sets (connect their components).
     *   - At the end, DSU.count equals the number of connected components.
     *
     * Complexity:
     *   Roughly O((n + m) α(n)) where m = edges.length and α is very small.
     */
    const d = new DSU(n);
    for (const [u, v] of edges) {
        d.unite(u, v);
    }
    return d.count;
}

// ===============================
// Test Harness for countComponents
// ===============================

function runTest(name, n, edges, expected) {
    /*
     * Runs a single test case and prints:
     *   - the test name
     *   - input n and edge list
     *   - computed number of connected components
     *   - expected answer for verification
     */
    const result = countComponents(n, edges);
    console.log(name);
    console.log("n =", n, "edges =", edges);
    console.log("Connected components =", result, "(expected", expected + ")");
    console.log();
}

// ===============================
// Test 1: Single connected chain (0-1-2-3)
// ===============================
{
    /*
     * Graph:
     *   0 - 1 - 2 - 3
     * All nodes connected → exactly 1 component.
     */
    const n = 4;
    const edges = [
        [0, 1],
        [1, 2],
        [2, 3]
    ];
    runTest("Test 1: Single connected component (chain)", n, edges, 1);
}

// ===============================
// Test 2: Two components + isolated node
// ===============================
{
    /*
     * Graph:
     *   Component A: 0 - 1 - 2
     *   Component B: 3 - 4
     *   Node 5 isolated
     * Total components: 3
     */
    // Component A: 0-1-2
    // Component B: 3-4
    // Node 5 isolated → own component
    const n = 6;
    const edges = [
        [0, 1],
        [1, 2],
        [3, 4]
    ];
    runTest("Test 2: Two components plus isolated node", n, edges, 3);
}

// ===============================
// Test 3: No edges at all
// ===============================
{
    /*
     * Graph:
     *   n = 5, edges = []
     * Every node is isolated → 5 components.
     */
    const n = 5;
    const edges = [];
    runTest("Test 3: No edges (all isolated)", n, edges, 5);
}

// ===============================
// Test 4: Fully connected (redundant edges)
// ===============================
{
    /*
     * Graph:
     *   0 - 1 - 2 - 3 - 4 plus an extra edge 1 - 3
     * Still all nodes are connected → 1 component.
     * The edge [1,3] is redundant (it adds an alternate route / cycle).
     */
    // All 5 nodes should collapse into one component
    const n = 5;
    const edges = [
        [0, 1],
        [1, 2],
        [2, 3],
        [3, 4],
        [1, 3] // redundant
    ];
    runTest("Test 4: Fully connected with redundant edges", n, edges, 1);
}

File: code_samples/section11/problems/problem11_5/problem11_5.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
 * DSU (Disjoint Set Union) / Union-Find
 *
 * Purpose:
 *   Maintain a partition of {0..n-1} into disjoint sets supporting:
 *     - find(x):    return a representative (root) for x's set
 *     - unite(a,b): merge the sets containing a and b
 *
 * Heuristics used:
 *   - Path compression in find(): flattens trees to speed up future queries
 *   - Union by rank in unite(): attaches smaller-rank tree under larger-rank tree
 *
 * Extra field:
 *   - count: tracks the current number of connected components (disjoint sets).
 *            Starts at n and decrements on each successful merge.
 */
struct DSU {
    vector<int> parent; // parent[i] is i's parent; roots have parent[i] == i
    vector<int> rnk;    // rank heuristic (approx. tree height)
    int count;          // number of disjoint sets currently present

    DSU(int n) : parent(n), rnk(n, 0), count(n) {
        /*
         * Initialize DSU with n singleton sets:
         *   - each node is its own parent (root)
         *   - all ranks start at 0
         *   - component count starts at n
         */
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    int find(int x) {
        /*
         * Find the representative (root) of x's set.
         *
         * Path compression:
         *   If x is not a root, recursively find the root and rewrite parent[x]
         *   to point directly to that root. This greatly improves performance.
         */
        if (parent[x] != x)
            parent[x] = find(parent[x]);
        return parent[x];
    }

    void unite(int a, int b) {
        /*
         * Merge the sets containing a and b.
         *
         * Steps:
         *   1) Find roots ra and rb.
         *   2) If they are already equal, a and b are already connected; do nothing.
         *   3) Otherwise, attach the lower-rank root under the higher-rank root.
         *      If ranks are equal, choose one as new root and increment its rank.
         *   4) Decrement count because two components merged into one.
         */
        int ra = find(a), rb = find(b);
        if (ra == rb) return; // already in the same set; component count unchanged

        if (rnk[ra] < rnk[rb]) {
            parent[ra] = rb;  // attach ra under rb
        } else if (rnk[ra] > rnk[rb]) {
            parent[rb] = ra;  // attach rb under ra
        } else {
            // equal ranks: pick ra as root and bump its rank
            parent[rb] = ra;
            rnk[ra]++;
        }
        count--; // successful union reduces number of components by 1
    }
};

int countComponents(int n, const vector<pair<int,int>>& edges) {
    /*
     * Counts connected components in an undirected graph on nodes 0..n-1,
     * using DSU.
     *
     * Parameters:
     *   n     : number of nodes
     *   edges : list of undirected edges (u, v)
     *
     * Approach:
     *   - Start with n singleton components.
     *   - For each edge (u, v), union(u, v).
     *   - The DSU's 'count' ends up as the number of components.
     *
     * Assumptions:
     *   - Node indices in edges are valid: 0 <= u,v < n.
     */
    DSU d(n);
    for (auto &e : edges)
        d.unite(e.first, e.second);
    return d.count;
}

// ==================================
// Test Helper
// ==================================
void runTest(const string& name, int n, const vector<pair<int,int>>& edges, int expected) {
    /*
     * Runs a single test case:
     *   - Prints the test name and the edge list
     *   - Computes component count via countComponents()
     *   - Prints the result alongside an expected value
     */
    int result = countComponents(n, edges);

    cout << name << "\n";
    cout << "n = " << n << ", edges = { ";
    for (auto& e : edges) cout << "(" << e.first << "," << e.second << ") ";
    cout << "}\n";

    cout << "Connected components = " << result
         << " (expected " << expected << ")\n\n";
}

// ==================================
// Main test harness
// ==================================
int main() {

    // Test 1: Single connected component (chain)
    {
        /*
         * Graph: 0--1--2--3
         * All nodes are connected in one component.
         * Expected components: 1
         */
        int n = 4;
        vector<pair<int,int>> edges = {
            {0,1}, {1,2}, {2,3}
        };
        runTest("Test 1: Single component chain", n, edges, 1);
    }

    // Test 2: Two components + isolated node
    {
        /*
         * Graph:
         *   Component A: 0--1--2
         *   Component B: 3--4
         *   Node 5 is isolated
         * Expected components: 3
         */
        int n = 6;
        vector<pair<int,int>> edges = {
            {0,1}, {1,2},  // component 1
            {3,4}          // component 2
            // 5 isolated → component 3
        };
        runTest("Test 2: Two components plus isolated", n, edges, 3);
    }

    // Test 3: No edges
    {
        /*
         * Graph: n=5, no edges.
         * Every node is isolated, so each is its own component.
         * Expected components: 5
         */
        int n = 5;
        vector<pair<int,int>> edges; // empty
        runTest("Test 3: No edges (all isolated)", n, edges, 5);
    }

    // Test 4: Fully connected via redundant edges
    {
        /*
         * Graph:
         *   0--1--2--3--4 plus an extra edge 1--3
         * Still a single connected component. Extra edges are redundant.
         * Expected components: 1
         */
        int n = 5;
        vector<pair<int,int>> edges = {
            {0,1}, {1,2}, {2,3}, {3,4}, {1,3}
        };
        runTest("Test 4: Fully connected with redundant edges", n, edges, 1);
    }

    return 0;
}
